src/small_array.jl

```diff
@@ -0,0 +1,133 @@
+"""
+    $(TYPEDEF)
+
+A mutable resizeable 3-length vector to implement non-allocating small vectors.
+
+$(TYPEDFIELDS)
+"""
+mutable struct Backing{T} <: AbstractVector{T}
+    """
+    Length of the buffer.
+    """
+    len::Int
+    x1::T
+    x2::T
+    x3::T
+
+    Backing{T}() where {T} = new{T}(0)
+    Backing{T}(x1) where {T} = new{T}(1, x1)
+    Backing{T}(x1, x2) where {T} = new{T}(2, x1, x2)
+    Backing{T}(x1, x2, x3) where {T} = new{T}(3, x1, x2, x3)
+end
+
+Base.size(x::Backing) = (x.len,)
+Base.isempty(x::Backing) = x.len == 0
+
+"""
+    $(TYPEDSIGNATURES)
+
+Value to use when removing an element from a `Backing`. Used so stored entries can be
+GC'ed when removed.
+"""
+defaultval(::Type{T}) where {T <: Number} = zero(T)
+defaultval(::Type{Any}) = nothing
+
+function Base.getindex(x::Backing, i::Int)
+    @boundscheck 1 <= i <= x.len
+    if i == 1
+        x.x1
+    elseif i == 2
+        x.x2
+    elseif i == 3
+        x.x3
+    end
+end
+
+function Base.setindex!(x::Backing, v, i::Int)
+    @boundscheck 1 <= i <= x.len
+    if i == 1
+        setfield!(x, :x1, v)
+    elseif i == 2
+        setfield!(x, :x2, v)
+    elseif i == 3
+        setfield!(x, :x3, v)
+    end
+end
+
+function Base.push!(x::Backing, v)
+    x.len < 3 || throw(ArgumentError("`Backing` is full"))
+    x.len += 1
+    x[x.len] = v
+end
+
+function Base.pop!(x::Backing{T}, v) where {T}
```

Comment:
```suggestion
function Base.pop!(x::Backing{T}) where {T}
```
The 2nd argument does not seem to be necessary.

Reply:
Thanks!